recipes/dcase2021_task4_baseline/finetune_on_separated.py

```diff
@@ -0,0 +1,364 @@
+import argparse
+from copy import deepcopy
+import numpy as np
+import os
+import pandas as pd
+import random
+import torch
+import yaml
+
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks import EarlyStopping, ModelCheckpoint
+from pytorch_lightning.loggers import TensorBoardLogger
+
+from desed_task.dataio import ConcatDatasetBatchSampler
+from desed_task.dataio.datasets import StronglyAnnotatedSet, UnlabelledSet, WeakSet
+from desed_task.nnet.CRNN import CRNN
+from desed_task.utils.encoder import ManyHotEncoder
+from desed_task.utils.schedulers import ExponentialWarmup
+
+from local.classes_dict import classes_labels
+from local.sepsed_trainer import SEPSEDTask4_2021
+from local.sed_trainer import SEDTask4_2021
+
+
+class EnsembleModel(torch.nn.Module):
+    def __init__(self, sed_model):
+        super(EnsembleModel, self).__init__()
+        self.multisrc_model = sed_model
+        self.monaural_model = deepcopy(sed_model)
+        self.q = torch.nn.Parameter(torch.rand((1)))
+
+    def forward(self, x, n_src, nosep):
+        strong, weak = self.multisrc_model(x)
+        _, clss, frames = strong.shape
+        strong = strong.reshape(-1, n_src, clss, frames)
+        weak = weak.reshape(-1, n_src, clss)
+
+        strong = torch.clamp(torch.sum(strong, 1), max=1)
+        weak = torch.clamp(torch.sum(weak, 1), max=1)
+
+        with torch.no_grad():
+            strong_nosep, weak_nosep = self.monaural_model(nosep)
+
+        strong = strong_nosep * self.q + strong * (1 - self.q)
+        weak = weak_nosep * self.q + weak * (1 - self.q)
```

Comment:
Interesting, you mix the predictions with the mixture model here depending on a random number q, right ? 
This is not for the code, but I would like to know why q is a random value (in the __init__, so will be fix during the all process ?) ?

Reply:
It is learned the parameter is added to the optimizer

Reply:
Ah, that's nice very then, I didn't notice that !!